lib/puppetserver/ca/sign_action.rb

```diff
@@ -0,0 +1,252 @@
+          @logger.err 'Error'
+          @logger.err "No waiting certificate requests to sign"
+          return nil
+        end
+
+        return requestedCertnames
+      end
+
+      def get_all_pending_certs
+        result = get_all_certs
+        if result
+          select_pending_certs(result)
+        end
+      end
+
+      def sign_requested_certs(certnames)
+        exit_code = 0
+        certnames.each do |certname|
+          result = sign_cert(certname)
+
+          if result.is_a? Net::HTTPNoContent
+            @logger.inform "Signed Certificate for #{certname}"
+          elsif result.is_a? Net::HTTPNotFound
+            @logger.err 'Error'
+            @logger.err "    Could not find certificate for #{certname}"
+            exit_code = 1
+          else
+            @logger.err 'Error - gross debug output:'
+            @logger.err "    #{result.inspect}"
+            exit_code = 1
+          end
+        end
+        return exit_code
+      end
+
+      def store
+        unless @store
+          store = OpenSSL::X509::Store.new
+          store.add_file(File.expand_path('~/.puppetlabs/etc/puppet/ssl/certs/ca.pem'))
+          store.add_file(File.expand_path('~/.puppetlabs/etc/puppet/ssl/certs/localhost.pem'))
+          store.add_file(File.expand_path('~/.puppetlabs/etc/puppet/ssl/crl.pem'))
+          @store = store
+        end
+        return @store
+      end
+
+      def check_flag_usage(results,parser)
+        if results['certname'] && results['all']
+          @logger.err '--all and --certname cannot be used together'
+          @logger.err parser.help
+          return false
+        elsif !results['certname'] && !results['all']
+          @logger.err 'No arguments given'
+          @logger.err parser.help
+          return false
+        elsif results['certname'] && results['certname'].include?('--all')
+          @logger.err 'Cannot use --all with --certname. If you actually have a certificate request ' +
+                          'for a certifcate named --all, you need to use the HTTP API.'
+          return false
+        end
+        return true
+      end
+
+      def log_unknown_arg(unparsed,parser)
+        @logger.err 'Error:'
+        @logger.err 'Unknown arguments or flags:'
+        unparsed.each do |arg|
+          @logger.err "    #{arg}"
+        end
+
+        @logger.err ''
+        @logger.err parser.help
+      end
+
+      def log_invalid_certname()
+        @logger.err 'Error:'
+        @logger.err 'A certname to sign is required'
+      end
+
+      def check_unparsed(unparsed,parser)
+        unless unparsed.empty?
+          log_unknown_arg(unparsed,parser)
+          return false
+        end
+        return true
+      end
+
+      def valid?(invalid,unparsed,results,parser)
+        unless invalid.empty?
+          log_invalid_certname
+          return false
+        else
+           [check_flag_usage(results,parser), check_unparsed(unparsed,parser)].all?
```

Comment:
this is a little ugly, but the idea is that we want to return false if either ``check_flag_usage`` or ``check_unparsed`` is false, and we want to make sure to call both methods. We can't use ``and`` or ``&&``, as both use short circuit evaluation (``and`` just has lower precedence, it turns out).

Reply:
The more I think about this the more I think the UX benefit isn't worth the weird code. What do you think?

Reply:
I think I agree... I'm not sure the intent/necessity behind this logic is coming through at all. I could put up a second commit with a simpler if/elsif/elsif check so we can compare the two?